### PR TITLE
WIP 256 Bit wide instruction optimization for avx2 capable processors

### PR DESCRIPTION
**Pull request by tgruben**
Some Intel processors can process bit operations in parallel via its vector processing capabilities.  These features are not currently exposed to the Go runtime so I have produced access via the go assembler.

### CODE REVIEW

File: roaring/asm_stub.go [hunk 1]
```diff
@@ -0,0 +1,10 @@
+package roaring
```
Comment (travisturner) [inline roaring/asm_stub.go]: this file needs a license header

**Comment by travisturner** [comment]
I didn't review the assembly, but the rest of it seems reasonable to me (with the exception of my previous comments).

**Comment by jaffee** [comment]
is difference not implemented @tgruben ?

**Comment by tgruben** [comment]
yeah its AndN @jaffee

**Comment by jaffee** [comment]
oh right - for some reason, I thought that was intersectionCount

File: roaring/roaring.go [hunk 2]
```diff
@@ -2137,15 +2138,95 @@ func intersectArrayBitmap(a, b *Container) *Container {
...
+	// see https://go101.org/article/bounds-check-elimination.html
+	var (
+		ab = a[:bitmapN]
+		bb = b[:bitmapN]
+		ob = c[:bitmapN]
+		n  int
+	)
+	for i := 0; i < bitmapN; i++ {
+		v := ab[i] & bb[i]
+		ob[i] = v
```
Comment (jaffee) [inline roaring/roaring.go]: why not set ob[i] directly? `ob[i] = ab[i] & bb[i]`

Comment (tgruben) [reply roaring/roaring.go]: because i use the stored value to get the bitcount on the next line

**Comment by jaffee** [comment]
paging @TocarIP you may be interested in this PR. We discussed vector instructions briefly at gophercon. This seems to show that the 256 bit instructions don't give much of a performance boost for bitmap intersects (though please check out the benchmarks for yourself). Loop unrolling (in Go) gives a moderate boost (~10% on my machine), but it was a bit unpredictable. Fully unrolling the loop gives no performance advantage.

We'd certainly appreciate any input you have on any of this.

**Comment by jaffee** [comment]
For reference:
```
BenchmarkBitmap_GoAndNoBCE-8                      	 1000000	      1233 ns/op    Baseline naive implementation
BenchmarkBitmap_GOIntersect-8                     	 2000000	       934 ns/op    Eliminated bounds checks
BenchmarkBitmap_ASMIntersect-8                    	 2000000	       931 ns/op    Assembly implementation with 256-bit vector instructions
BenchmarkBitmap_GOIntersectUnroll4-8              	 2000000	       856 ns/op    Loop unrolled in Go 4 iterations
BenchmarkBitmap_GOIntersectUnroll8-8              	 2000000	       839 ns/op    Loop unrolled in Go 8 iterations
BenchmarkBitmap_GOIntersectUnroll16-8             	 2000000	       815 ns/op    Loop unrolled in Go 16 iterations
BenchmarkBitmap_GOIntersectUnroll1024-8           	 2000000	       926 ns/op    Loop removed completely
```

**Comment by TocarIP** [comment]
Usually vectorized popcount uses some form of vectorized bit tricks to keep e. g. 4 partial popcounts in one vector register, and reduces it into single value in the end.  https://arxiv.org/pdf/1611.07612.pdf does something similar.

**Comment by tgruben** [comment]
Closing this PR till vectorized popcnt is more generally available
my unroll script
`
print """TEXT ·asmAnd(SB), 7, $0
        MOVQ $0, SI
        MOVQ a_data+0(FP), BX   // BX = &a[0]
        MOVL a_len+8(FP), CX    // len(a)
        MOVQ b_data+24(FP), DX  // DX = &b[0]
        MOVQ c_data+48(FP), R15 // DX = &c[0]
        VPXORQ Z2,Z2,Z2"""

mes="""
VMOVDQA64 %d(BX), Z0
VMOVDQA64 %d(DX), Z1
VPANDQ   Z0, Z1, Z0
VMOVUPS Z0, %d(R15)
VPOPCNTQ Z0,Z0,Z0
VPADDQ Z0,Z2,Z2
"""
x=0
for i in range(0,8192,64):
   print mes%(i,i,i)
   x+=64

print """JNE  loop_begin3
        MOVQ SI, ·noname+72(FP)
        VZEROUPPER
        RET"""`

**Pull request closed without merging**